### DISCUSSION TITLE
Azure OpenAI Models returning 'Connection error' in latest release

### DISCUSSION THREAD

**Discussion by jrkropp** [danny-avila/LibreChat#1411] [Troubleshooting]
### What happened?

**Error message:** Something went wrong. Here's the specific error message we encountered: Error: Connection error.
![image](https://github.com/danny-avila/LibreChat/assets/132182384/ae920c82-b190-49bf-91fb-077adb72180c)

I ONLY receive the error starting with v0.6.5.  The issue only occurs with Azure OpenAI endpoint.  OpenAI endpoint works fine.

🟥danny-avila/librechat-dev:8d563d61f1ed12c88a279b2f16a4782ef93cc49f has the issue.
✅danny-avila/librechat-dev:c9d3e0ab6a42bafee82ef1f6cd5174a9561e21c5 works as expected.
[https://github.com/danny-avila/LibreChat/pkgs/container/librechat-dev/versions](url)

**Logs:**
tion node api/server/index.js
2023-12-20T20:23:47.033496681Z 
2023-12-20T20:23:48.624310920Z 2023-12-20 20:23:48 info: `REDIS_URI` not provided, or `USE_REDIS` not set. Redis module will not be initialized.
2023-12-20T20:23:50.799587244Z 2023-12-20 20:23:50 info: Connected to MongoDB
2023-12-20T20:23:50.802386758Z Warning: connect.session() MemoryStore is not
2023-12-20T20:23:50.802398270Z designed for a production environment, as it will leak
2023-12-20T20:23:50.802401095Z memory, and will not scale past a single process.
2023-12-20T20:23:50.813205605Z 2023-12-20 20:23:50 info: Server listening on all interfaces at port 3080. Use http://localhost:3080 to access it
2023-12-20T20:24:57.430347322Z 2023-12-20 20:24:57 warn: [OpenAIClient.chatCompletion][stream] API error
2023-12-20T20:24:57.431057257Z 2023-12-20 20:24:57 warn: [OpenAIClient.chatCompletion][finalChatCompletion] API error
2023-12-20T20:24:57.432055240Z 2023-12-20 20:24:57 error: [OpenAIClient.chatCompletion] Unhandled error type Error: Connection error.
2023-12-20T20:24:57.432472313Z 2023-12-20 20:24:57 error: [handleAbortError] response error and aborting request Error: Connection error.

![image](https://github.com/danny-avila/LibreChat/assets/132182384/f0b7720a-5aa7-4769-ae8c-0b18d33bc4c2)
![image](https://github.com/danny-avila/LibreChat/assets/132182384/10d2edc0-35f7-44a8-b9b7-9449471fa6cc)

AZURE_OPENAI_API_DEPLOYMENT_NAME: gpt-4
AZURE_OPENAI_API_VERSION: 2023-07-01-preview
AZURE_OPENAI_MODELS: gpt-4,gpt-4-0613,gpt-4-1106-preview
AZURE_USE_MODEL_AS_DEPLOYMENT_NAME: true
ENDPOINTS: azureOpenAI,openAI

Note: Apologies if I didn't provide sufficient info or in the wrong way.  This is the first issue I've reported on Github so learning the ropes.

### Steps to Reproduce

1. Deploy latest version (v0.6.5).  Or danny-avila/librechat-dev:8d563d61f1ed12c88a279b2f16a4782ef93cc49f
2. Configure for Azure OpenAI
AZURE_OPENAI_API_DEPLOYMENT_NAME: gpt-4
AZURE_OPENAI_API_VERSION: 2023-07-01-preview
AZURE_OPENAI_MODELS: gpt-4,gpt-4-0613,gpt-4-1106-preview
AZURE_USE_MODEL_AS_DEPLOYMENT_NAME: true
ENDPOINTS: azureOpenAI,openAI

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Thanks for your report. Azure was working great for me on the latest commit, otherwise I would not have cut a release. Check the error logs in /api/logs with error in the file name. 

Also the Azure setup guide was recently updated. I basically ran through a clean setup with no issues. Please review the updated instructions in case anything sticks out: https://docs.librechat.ai/install/ai_setup.html#azure-openai

**Comment by danny-avila** [comment]
The major change was that the official SDK for OpenAI via node.js is now used,but this is not really a breaking change. Do you have all the required variables set?

**Reply by jrkropp** [reply]
Thanks for responding. I'm running LibreChat as an 'Azure Container App' so I don't have easy access to the underlying files/logs.  Just what's in the log stream.  I'll try running it locally and see if I can reproduce it.

I've double-checked the environment variables and they seem to be correctly set (copied below).  Strange that its only the latest version that isn't working.  Reverting to the previous version without changing any variables fixes the issue.

**AZURE_API_KEY:** ********
**AZURE_OPENAI_API_INSTANCE_NAME:** ********
**AZURE_OPENAI_API_DEPLOYMENT_NAME:** gpt-4
**AZURE_OPENAI_API_VERSION:** 2023-07-01-preview _(I've tried all the versions)_
**AZURE_OPENAI_MODELS:** gpt-4,gpt-4-vision-preview,gpt-4-1106-preview,gpt-4-0613,gpt-4-32k,gpt-4-32k-0613,gpt-35-turbo,gpt-35-turbo-1106,gpt-35-turbo-16k
**AZURE_USE_MODEL_AS_DEPLOYMENT_NAME:** true
**ENDPOINTS:** azureOpenAI,openAI

![image](https://github.com/danny-avila/LibreChat/assets/132182384/766e9e5e-bf7a-401e-91a1-dc10aa7afa39)

I have configured Azure AD for OpenID authentication as outlined in the official documentation. The application connects to an existing MongoDB instance that already contains data from previous usage of LibreChat. I'm not certain if the pre-existing data is causing the issue. It's worth noting that coinciding with the update to the latest version, all my chat history and prompts were lost, though this may have been indirectly caused when I was troubleshooting this issue.  In an attempt to resolve the issue, I deleted and then recreated my Azure OpenAI instance/models, and recreated using the recommended naming in the updated guide. I left the MongoDB instance and other configurations untouched, except for the Azure model names and OpenAI variables.  Not sure why the history/presets where lost.

Also tried with single model/deployment name and got the same error.  Also tried a few different Azure OpenAI regions and got the same thing.

If I'm the only one having this issue it must be me. My next steps are:
- Run locally to see if I can reproduce.  If not, it might be an incompatibility with Azure Container apps.
- Wipe my Mongo database and start fresh.
- Remove my Azure AD OpenID auth and see if it works.  It's a longshot, but maybe that's the only thing different between my and your instance.

**Reply by jrkropp** [reply]
**Update:** the lost of my Mongo database was confirmed my fault.  I forgot I correct my Mongo URI when rebuilding.

**Before:** mongodb+srv://*************.mongodb.net/?retryWrites=true
**After:** mongodb+srv://*************.mongodb.net/**LibreChat**?retryWrites=true

The addition/correction of adding LibreChat to the end caused it to create a new collection called LibreChat, which make sense.  Before that it defaulted to creating a collection called test.

**Reply by danny-avila** [reply]
Thanks. Without those logs, I can’t easily debug this. Maybe this is relevant? https://techcommunity.microsoft.com/t5/apps-on-azure-blog/downloading-files-from-azure-container-app-to-your-local-machine/ba-p/3774872

I’ve never run this as an azure container app so any instructions for that to help me reproduce would be appreciated

**Reply by jrkropp** [reply]
Thanks!  Don't worry about troubleshooting further if I'm the only one reporting this issue.  I'll spend some time next week looking deeper into it.

Setting up LibreChat via an 'Azure Container App' instructions below (different than Azure Container Instance).  It's probably the easiest and most cost effective option for hosting LibreChat in Azure.  5min setup and only costs a few dollars a month to host, has built-in scaling options, built-in free custom managed SSL cert, secure secret management, etc...

1) Go to the Azure portal at https://portal.azure.com/ and use the search bar to find 'Container Apps'. Click 'Create'.
![image](https://github.com/danny-avila/LibreChat/assets/132182384/6b55cdb2-15c0-4696-b19c-863b657c36b8)

2) Fill in your Azure subscription details, resource group, container app name, and select the appropriate region.
![image](https://github.com/danny-avila/LibreChat/assets/132182384/ce65baae-e512-49e5-9ab1-e42ded98f8ab)

3) If you haven't set up a container environment yet, click 'Create'. Choose the 'Consumption' plan and proceed with the default settings.
![image](https://github.com/danny-avila/LibreChat/assets/132182384/d9f3a826-e78e-4a95-a6f1-198c12b7a710)

4) In the 'Container' tab, enter a container name, choose 'Docker Hub or other registries' for the image source, and input the following details:
- Registry login server: ghcr.io
- Image and tag: danny-avila/librechat:latest
- Environment variables: Populate these based on the .env.example file, e.g., APP_TITLE, HOST, PORT=3080, etc.....

![image](https://github.com/danny-avila/LibreChat/assets/132182384/01b6b359-e9d0-4646-9145-22ee641c8169)

5) Skip the 'Bindings' tab and leave it empty.
![image](https://github.com/danny-avila/LibreChat/assets/132182384/63ff0e72-7313-4b5c-b984-fd0a467a9bed)

6) In the 'Ingress' tab, turn it on and select 'Accepting traffic from anywhere'. This can be restricted to specific IPs later if needed. Input the target port mentioned in your environment variables (default is 3080).
![image](https://github.com/danny-avila/LibreChat/assets/132182384/24e5ba48-108d-4256-b7ea-5968434a8e7f)

7) Click 'Create' to finish the setup.

Once your container app is live, you can assign a custom domain or use the one provided. Microsoft offers a free managed SSL cert option and configure bindings for you for easy https:// setup.

Cost is based on consumption.  I pay ~$5-15/month but you could get this down further by adjusting scaling to zero so it's not running 24'7.

For added security, limit access to certain IPs via the 'Ingress' settings or set up an additional layer of protection via the authentication pane, although this may be unnecessary if you're already using Azure AD OpenId auth within the app.  But gives corporate environments extra peace of mind that the application can't be accessed without first authenticating. Adjust scaling options as needed in the 'Scale and replicas' section.

**Reply by danny-avila** [reply]
> Image and tag: danny-avila/librechat:latest

try the other image, whose build steps you can see in Dockerfile.multi and setup you can see in deploy-compose.yml, both in root directory.

`danny-avila/librechat-api:latest`

**Comment by danny-avila** [comment]
I can confirm I have no issues with Azure using user_provided credentials and credentials stored in env file.

here are my variables:

```bash
AZURE_API_KEY=REDACTED
AZURE_OPENAI_API_INSTANCE_NAME=REDACTED
AZURE_OPENAI_API_DEPLOYMENT_NAME=gpt-4 # specified but not necessary as using model name as deployment name
AZURE_OPENAI_API_VERSION=2023-12-01-preview
AZURE_OPENAI_MODELS=gpt-4,gpt-4-vision-preview,gpt-4-1106-preview,gpt-4-0613,gpt-4-32k,gpt-4-32k-0613,gpt-3.5-turbo,gpt-3.5-turbo-1106,gpt-3.5-turbo-16k
AZURE_USE_MODEL_AS_DEPLOYMENT_NAME=TRUE
```



![IMG_7750](https://github.com/danny-avila/LibreChat/assets/110412045/90fe093d-bd3d-4ee4-b9be-f39f97483a55)

![IMG_7749](https://github.com/danny-avila/LibreChat/assets/110412045/16f41d82-bf36-4e58-860b-afd19b292cb1)

it would help to see your error logs. Not sure how it works with azure containers, since mounting volumes/storage might look different there, but here’s how it would with docker.

`docker ps` - run this to get all running containers, we need the container ID of the LibreChat API container

`docker exec -it <container_id> /bin/sh` - to enter shell of container

`cd /app/api/logs` - enter logs directory

`ls` - list all files in current directory, we need the most recent error logs

`cat filename.log` - view the log file

Or `tail -n 100 filename.log` - view the last few lines

**Reply by danny-avila** [reply]
What it says for you here might be useful to see (from the debug logs, not error)

```
2023-12-25 13:29:35 debug: [OpenAIClient] chatCompletion
{
        baseURL: "https://<INSTANCE-NAME>.openai.azure.com/openai/deployments/gpt-35-turbo/chat/completions?api-versi... [truncated]",
```
